File: Basic_matting/solve.py
import numpy as np


def solve(mu_F, Sigma_F, mu_B, Sigma_B, C, sigma_C, alpha_init, maxIter, minLike):
    
    I = np.eye(3)
    max_F = np.zeros(3)
    max_B = np.zeros(3)
    max_alpha = 0
    maxlike = - np.inf
    invsgma2 = 1/sigma_C**2
    for i in range(mu_F.shape[0]):
        mu_Fi = mu_F[i]
        Inv_variance_Fi = np.linalg.inv(Sigma_F[i])
        for j in range(mu_B.shape[0]):
            mu_Bj = mu_B[j]
            Inv_Variance_Bj = np.linalg.inv(Sigma_B[j])

            alpha = alpha_init
            myiter = 1
            lastLike = -1.7977e+308
            while True:
                # solve for F,B
                A11 = Inv_variance_Fi + I*alpha**2 * invsgma2
                A12 = I*alpha*(1-alpha) * invsgma2
                A22 = Inv_Variance_Bj+I*(1-alpha)**2 * invsgma2
                A = np.vstack((np.hstack((A11, A12)), np.hstack((A12, A22))))
                b1 = Inv_variance_Fi @ mu_Fi + C*(alpha) * invsgma2
                b2 = Inv_Variance_Bj @ mu_Bj + C*(1-alpha) * invsgma2
                b = np.atleast_2d(np.concatenate((b1, b2))).T

                X = np.linalg.solve(A, b)
                F = np.maximum(0, np.minimum(1, X[0:3]))
                B = np.maximum(0, np.minimum(1, X[3:6]))
                # solve for alpha

                alpha = np.maximum(0, np.minimum(1, ((np.atleast_2d(C).T-B).T @ (F-B))/np.sum((F-B)**2)))[0,0]
                # # calculate likelihood
                L_C = - np.sum((np.atleast_2d(C).T -alpha*F-(1-alpha)*B)**2) * invsgma2
                L_F = (- ((F- np.atleast_2d(mu_Fi).T).T @ Inv_variance_Fi @ (F-np.atleast_2d(mu_Fi).T))/2)[0,0]
                L_B = (- ((B- np.atleast_2d(mu_Bj).T).T @ Inv_Variance_Bj @ (B-np.atleast_2d(mu_Bj).T))/2)[0,0]
                like = (L_C + L_F + L_B)
                #like = 0

                if like > maxlike:
                    max_alpha = alpha
                    maxlike = like
                    max_F = F.ravel()
                    c = B.ravel()

                if myiter >= maxIter or abs(like-lastLike) <= minLike:
                    break

                lastLike = like
                myiter += 1
    return max_F, max_B, max_alpha

File: solve.m
function [F, B, alpha] = solve(mean_F, Sigma_F, mean_B, Sigma_B, C, sigma_C, alpha_init, maxIter, minLike)

    I = eye(3);
    maxLike = -realmax;
    
    for i = 1:size(mean_F, 2)
        mean_Fi = mean_F(:, i);
        invSigma_Fi = inv(Sigma_F(:, :, i));
                
        for j = 1:size(mean_B, 2)
            mubi = mean_B(:, j);
            invSigmabi = inv(Sigma_B(:, :, j));
            
            alpha = alpha_init;
            iter = 1;
            lastLike = -realmax;
            
            while (1)
                
                    % solve for F,B
					alpha_sq_inv = alpha^2/sigma_C^2;
					one_alpha_sq_inv = (1-alpha)^2/sigma_C^2;
					alpha_one_alpha_inv = alpha*(1-alpha)/sigma_C^2;
					A = [invSigma_Fi+I*alpha_sq_inv, I*alpha_one_alpha_inv;                  I*alpha_one_alpha_inv, invSigmabi+I*one_alpha_sq_inv];
					b = [invSigma_Fi*mean_Fi + C*(alpha/sigma_C^2);                  invSigmabi*mubi + C*((1-alpha)/sigma_C^2)];
					X = A\b;
					F = max(0,min(1,X(1:3)));
					B = max(0,min(1,X(4:6)));
					
					% solve for alpha
					F_minus_B = F - B;
					alpha_num = (C-B)'*F_minus_B;
					alpha_den = sum(F_minus_B.^2);
					alpha = max(0,min(1,alpha_num/alpha_den));
					
					% calculate likelihood
					C_minus_alphaF_minus_1minusalphaB = C - alpha*F - (1-alpha)*B;
					L_C = -sum(C_minus_alphaF_minus_1minusalphaB.^2)/sigma_C;
					L_F = -(F-mean_Fi)'*invSigma_Fi*(F-mean_Fi)/2;
					L_B = -(B-mubi)'*invSigmabi*(B-mubi)/2;
					like = L_C + L_F + L_B;
					
					if iter >= maxIter || abs(like-lastLike) <= minLike
						break;
					end
					
					lastLike = like;
					iter = iter + 1;
            end
            
            % keep only the maximum likelihood estimates
            if like > maxLike
                maxLike = like;
                maxF = F;
                maxB = B;
                maxAlpha = alpha;
            end
        end
    end
    
    % return maximum likelihood estimates
    F = maxF;
    B = maxB;
    alpha = maxAlpha;
    
end
